library/core/src/iter/adapters/zip.rs

```diff
@@ -171,26 +298,53 @@ macro_rules! zip_impl_general_defaults {
             // and doesn’t call `next_back` too often, so this implementation is safe in
             // the `TrustedRandomAccessNoCoerce` specialization
 
-            let a_sz = self.a.len();
-            let b_sz = self.b.len();
-            if a_sz != b_sz {
-                // Adjust a, b to equal length
-                if a_sz > b_sz {
-                    for _ in 0..a_sz - b_sz {
-                        self.a.next_back();
-                    }
-                } else {
-                    for _ in 0..b_sz - a_sz {
-                        self.b.next_back();
-                    }
-                }
-            }
+            self.adjust_back();
             match (self.a.next_back(), self.b.next_back()) {
                 (Some(x), Some(y)) => Some((x, y)),
                 (None, None) => None,
                 _ => unreachable!(),
             }
         }
+
+        #[inline]
+        default fn fold<T, F>(mut self, init: T, f: F) -> T
+        where
+            F: FnMut(T, Self::Item) -> T,
+        {
+            ZipImpl::try_fold(&mut self, init, ok(f)).unwrap()
```

Comment:
Yup, as [the8472](https://github.com/the8472) said,
```suggestion
            ZipImpl::try_fold(&mut self, init, NeverShortCircuit::wrap_mut_2(f)).0
```
so that neither the compiler nor the reader needs to spend time proving that it's not `Err`.